packages/frontend/core/src/modules/docs-search/entities/docs-indexer.ts

```diff
@@ -0,0 +1,299 @@
+import { DebugLogger } from '@affine/debug';
+import type { Job, WorkspaceService } from '@toeverything/infra';
+import {
+  Document,
+  Entity,
+  IndexedDBIndexStorage,
+  IndexedDBJobQueue,
+  JobRunner,
+} from '@toeverything/infra';
+import { difference } from 'lodash-es';
+import type { Array as YArray, Map as YMap } from 'yjs';
+import { applyUpdate, Doc as YDoc } from 'yjs';
+
+import { blockIndexSchema, docIndexSchema } from '../schema';
+
+const logger = new DebugLogger('crawler');
+
+interface IndexerJobPayload {
+  docId: string;
+}
+
+export class DocsIndexer extends Entity {
+  private readonly jobQueue = new IndexedDBJobQueue<IndexerJobPayload>(
+    'jq:' + this.workspaceService.workspace.id
+  );
+
+  private readonly runner = new JobRunner(
+    this.jobQueue,
+    (jobs, signal) => this.execJob(jobs, signal),
+    () => new Promise<void>(resolve => requestIdleCallback(() => resolve()))
+  );
+
+  private readonly indexStorage = new IndexedDBIndexStorage(
+    'idx:' + this.workspaceService.workspace.id
+  );
+
+  readonly docIndex = this.indexStorage.getIndex('doc', docIndexSchema);
+
+  readonly blockIndex = this.indexStorage.getIndex('block', blockIndexSchema);
+
+  private readonly workspaceEngine = this.workspaceService.workspace.engine;
+
+  private readonly workspaceId = this.workspaceService.workspace.id;
+
+  constructor(private readonly workspaceService: WorkspaceService) {
+    super();
+  }
+
+  setupListener() {
+    this.workspaceEngine.doc.storage.eventBus.on(event => {
+      if (event.clientId === this.workspaceEngine.doc.clientId) {
+        const docId = event.docId;
+
+        this.jobQueue
+          .enqueue([
+            {
+              batchKey: docId,
+              payload: { docId },
+            },
+          ])
+          .catch(err => {
+            console.error('Error enqueueing job', err);
+          });
+      }
+    });
+  }
+
+  async execJob(jobs: Job<IndexerJobPayload>[], _signal: AbortSignal) {
+    if (jobs.length === 0) {
+      return;
+    }
+
+    // jobs should have the same docId, so we just pick the first one
+    const docId = jobs[0].payload.docId;
+
+    logger.debug('Start crawling job for docId:', docId);
+
+    if (docId) {
+      if (docId === this.workspaceId) {
+        await this.crawlingRootDocData();
+      } else {
+        await this.crawlingDocData(docId);
+      }
+    }
+  }
+
+  startCrawling() {
+    this.runner.start();
+    this.jobQueue
+      .enqueue([
+        {
+          batchKey: this.workspaceId,
+          payload: { docId: this.workspaceId },
+        },
+      ])
+      .catch(err => {
+        console.error('Error enqueueing job', err);
+      });
+  }
+
+  async crawlingDocData(docId: string) {
+    const rootDocBuffer =
+      await this.workspaceEngine.doc.storage.loadDocFromLocal(this.workspaceId);
+
+    if (!rootDocBuffer) {
+      return;
+    }
+
+    const yRootDoc = new YDoc();
```

Comment:
ydoc instance again?

Reply:
What your mean? We cannot get the data without creating an instance.